sound/soc/sof/intel/hda.c

```diff
@@ -681,6 +681,19 @@ void hda_ipc4_dump(struct snd_sof_dev *sdev)
 		hipci, hipcie, hipct, hipcte, hipcctl);
 }
 
+bool hda_ipc4_can_send_message(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	u32 val;
+
+	val = snd_sof_dsp_read(sdev, HDA_DSP_BAR, chip->ipc_req);
```

Comment:
is there no internal IPC state indicator to check that we're busy after having sent a request and expecting a response?

Reply:
This is the hardware handshake between the two processors.
Initially the assumption was that when the DSP sent a reply it is free to receive a message. This turned out to be not the case: the SOF firmware is acking the message _after_ it sent the reply message, which is fine most of the time... But there are cases when the time between the reply sending and acking of the received message (the one the reply was sent) is so long that the kernel _might_ send another message (without checking that the channel is free), the firmware get to the point to ack *the previous* message and it actually acks the current one, loosing it completely.

We are doing to be safe even if the DSP does something unexpected, but not totally invalid.